### ISSUE TITLE
IINA requests access to Apple Music

### ISSUE THREAD

**Issue by low-batt** [iina/iina#5305]
### IINA Version

1.3.5

### Release Channel

Release

### macOS Version

15.2

### Steps to Reproduce

On a Mac running macOS Sequoia…
- Under the  menu click on `System Settings…`
- In the `System Settings` panel click on `Privacy & Security`
- On the right side of the panel lick on `Media & Apple Music`
- If IINA is listed click on it
- Click on the `-` button
- If prompted for admin credentials, enter them
- Using IINA, play a video that is in the `Movies` directory under your home directory
- A pop-up appears indicating IINA would like to access Apple Music

### Expected Behavior

IINA does not request access to Apple Music.

### Actual Behavior

The first time a video is played that is in the `Movies` directory or a song is played that is in the `Music` directory macOS Sequoia will display the following alert:
![alert](https://github.com/user-attachments/assets/9b74c94d-0499-46d0-90a7-280d093db2bf)

This is a violation of the [Privacy](https://developer.apple.com/design/human-interface-guidelines/privacy) section of Apple's [Human Interface Guidelines](https://developer.apple.com/design/human-interface-guidelines/), specifically:
> **Write copy that clearly describes how your app uses the ability, data, or resource you’re requesting.** The standard alert displays your copy (called a purpose string or usage description string) after your app name and before the buttons people use to grant or deny their permission.

The above alert is missing the "purpose string". To supply such a string IINA would need to add a [NSAppleMusicUsageDescription](https://developer.apple.com/documentation/bundleresources/information-property-list/nsapplemusicusagedescription) key with a string value such as "This app requires Apple Music library access to play music tracks." to the `Info.plist` file.

_However_ the HIG also says:
> **Request permission only when your app clearly needs access to the data or resource.**

Since IINA does not support accessing Apple Music it is odd that it is requesting access.

### How often can you reproduce the issue

Always, under macOS Sequoia

### Required Actions

- [X] I tried my best to search from the existing issues (including closed issues) and didn't find duplicated issues.

### Optional Actions

- [X] I tested with the Nightly build and confirmed the issue is not fixed.
- [X] I tested with the latest version of mpv and confirmed mpv doesn't have this issue.

**Comment by svobs** [comment]
@low-batt I see you self-assigned this so I'm curious if you think you have a lead on a fix. I dug into a rabbit hole and found:
1. The culprit (for me) was the `AutoFileMatcher`, and oddly seemed to trigger the security prompt when the contents of subdirs inside the Movies folder _which are owned by Apple apps_ were accessed, not the folder itself. So the `Add files in the same folder automatically` didn't trigger it, because it only checked for files. The search for subtitles was the cause, because it tries to scan all the subdirs (& bundles, because they look like dirs) in that folder.
2. It seems like there might actually be a widespread misunderstanding about the purpose of the Movies folder. I myself thought it was intended as a sort of media folder, but I now think that Apple never intended it to be used for anything other than Apple files, starting with iMovie. This would be consistent with Apple's policy of naming things as simply as possible (e.g., iPhoto is now just called Photos). The prompt above pops up when Apple dirs inside it are accessed. But I have a few different directories in there created from shared albums and iMovie projects, with arbitrary names, and I can't find any way to query the OS to determine who "really" owns them other than to try and then get hit with the prompt.

**Comment by low-batt** [comment]
Yes, I'm working on this one although I just set it asside for the moment to help with Beta 1 release tasks. I could use an answer to the following question from someone who has access to Music or TV:

Are there media files in the directories owned by these services that are playable by IINA?

I'm expecting the answer is yes and there are times when IINA will need the permission controlled by this alert, just not in the example at hand.

As you noted the alert is being triggered by this call to [contentsOfDirectory(at:includingPropertiesForKeys:options:)](https://developer.apple.com/documentation/foundation/filemanager/1413768-contentsofdirectory) in AutoFileMatcher.getAllPossibleSubs:
https://github.com/iina/iina/blob/f03c0ff14557057221abb68d986de2e042d2fdcf/iina/AutoFileMatcher.swift#L108

In IINA's settings on the `Subtitle` tab under the `Auto Load` setting there is an `Advanced` disclosure triangle. Opening that exposes the `Also search subtitles in following directories` setting. By default that is set to `./*`. In the case at hand the video being played is in `~/Movies`. The alert is triggered by the call to `contentsOfDirectory` to get the contents of `~/Movies/TV`:
```text
low-batt@urp TV$ pwd
/Users/low-batt/Movies/TV
low-batt@urp TV$ ls -la
total 16
drwxr-xr-x   6 low-batt  staff   192 Dec 13 14:50 .
drwx------  74 low-batt  staff  2368 Nov 30 09:38 ..
-rw-r--r--@  1 low-batt  staff  6148 Dec 13 17:45 .DS_Store
drwxr-xr-x   4 low-batt  staff   128 Jul 28  2023 Media.localized
drwxr-xr-x   7 low-batt  staff   224 Dec 13 14:50 Previous Libraries.localized
drwxr-xr-x@  4 low-batt  staff   128 Sep 27 22:14 TV Library.tvlibrary
low-batt@urp TV$ 
```

I believe IINA will scan for subtitles even if `Auto load` is set to `Disabled`. That needs to be fixed.

I used `xattr` to look for extended file attributes would identify these files as belonging to these services. The file `TV Library.tvlibrary` is marked as belonging to TV. But I did not find anything on the `TV` directory that IINA could key off of to know these files are special.

The location where Music or TV stores files is not fixed and can be changed by settings, I've not tried it yet, but I'm expecting looking to see if `TV/TV Library.tvlibrary` will trigger the alert. If so, then I am expecting IINA will have to read the settings of those services to determine where these special directories are.

If there isn't a possibility of subtitle files being in these directories then IINA can skip them when searching for subtitles in subdirectories. If there is then we'd want a new setting to control whether IINA should scan into these directories.

I didn't find an official Apple statement on why this new alert was added. I read something about Apple having caught some apps inappropriately scanning these files for fingerprinting purposes.

As this alert is triggered during development whenever the executable file changes it is really irritating for developers. Definitely need a way to control this behavior.

Thoughts?

**Comment by svobs** [comment]
> As this alert is triggered during development whenever the executable file changes it is really irritating for developers. Definitely need a way to control this behavior.

Apparently this mechanism is called [TCC](https://www.angelystor.com/posts/macos_tcc/). I find the whole thing irritating, and I can't turn it off. So I have resorted to using a completely separate directory tree for my everyday use.

> I used xattr to look for extended file attributes would identify these files as belonging to these services. The file TV Library.tvlibrary is marked as belonging to TV. But I did not find anything on the TV directory that IINA could key off of to know these files are special.

These are the easy ones. The object `TV Library.tvlibrary` and others like it are actually a special Darwin/MacOS invention called a "bundle" (or sometimes referred to as a "package"). We need to filter them out because they will return true if asked if they are directories. But fortunately this is not hard to do. Bundles have a special "bundle bit" which is set, and this method can be used to quickly check it: 
`NSWorkspace.shared.isFilePackage(atPath: url.path)`.

Unfortunately we can still get a TCC challenge if we keep digging inside subdirectories of `Movies` which are owned by iMovie or other iApps (for example, deeper inside the `TV` directory), and I don't have a perfect solution for that. And the message text doesn't change; it still mentions Apple Music.

There is a system query which will return the location of the `Movies` dir:
`FileManager.default.urls(for: .moviesDirectory, in: .allDomainsMask)`
I haven't tried changing the location of my iMovie data, so I don't know if it would change what this returns. And unfortunately it's not helpful for identifying Apple-owned directories inside it; it only returns an array of a a single element, the Movies directory.

> I believe IINA will scan for subtitles even if Auto load is set to Disabled. That needs to be fixed.

Definitely agree with this. I would not have thought to look in the Subtitles settings for this one, or that it was doing such broad scanning for subtitles. I think:
1. All searches which return directories should be changed to check for, & filter out, bundles.
2. There should be a checkbox in front of `Also search subtitles in following directories` so that it can be disabled completely. (Also, this text should read "Also search _for_ subtitles in _the_ following directories"…)
3. We perhaps need a checkbox saying something like "Avoid folders which need special permission (Documents, Downloads, etc.)". We can use `FileManager` as above to get their locations and blacklist their subtrees from any filesystem scans.

**Comment by low-batt** [comment]
I tested IINA with code to read the TV settings and avoid accessing the configured directories containing the files that belong to TV ( `~/Movies/TV` by default). Didn't work because macOS is also now protecting the settings file of TV. Merely reading that property file triggers this same alert.

With respect to normal IINA users this issue is mostly just a concern with users asking why IINA needs access to Music as once a user allows access they will not be prompted again.

_However_ for developers and to some degree users that run Nightly builds this is a significant irritant as macOS will put up this prompt each time the IINA code signature changes. Looking like these users will have to change the `Also search subtitles in following directories` setting to be `.` instead of the default `./*` and maybe a list of the names of directories they keep subtitles in, something like `.:./subs/*`.

There are still problems that can be fixed. IINA should not be scanning when `Auto load` is set to `Disabled` as well as the problems with wording you pointed out.

We don't want to put too much work into this feature at this time. Instead we should be adding a UI to make it easy to switch to mpv's [sub-auto](https://mpv.io/manual/stable/#options-sub-auto) feature. There are open issues against IINA's automatic subtitle loading feature that use mpv's feature as a workaround.
